In [ ]:
!pip install -q ultralytics
from ultralytics import YOLO
model = YOLO("yolo11n.pt")
results = model.train(data="VisDrone.yaml", epochs=30, imgsz=640, batch=16, device=0, workers=2, project="runs_colab", name="baseline_yolo11n", pretrained=True, save=True, plots=True)

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 30.2 MB/s eta 0:00:00
Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
Ultralytics 8.4.38 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=VisDrone.yaml, degrees=0.0, deterministic=True, device=0, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=30, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, fre

The baseline YOLO11n model was trained for 30 epochs on the VisDrone dataset. The training process shows a steady decrease in loss values and a consistent increase in validation performance.

The final evaluation results are:
- mAP50-95: 0.158
- mAP50: 0.282
- Precision: 0.407
- Recall: 0.317

During training, the model demonstrates stable convergence behavior. The validation mAP steadily increases across epochs, indicating effective learning. There is no strong evidence of overfitting, as validation performance does not degrade in later epochs.

However, the overall performance is still relatively limited, especially for small object categories such as bicycle and tricycle. This is expected due to the challenging nature of the VisDrone dataset, which contains dense scenes and many small objects.

This baseline serves as a reference point for further experiments and model improvements.

In [ ]:
from ultralytics import YOLO

model = YOLO("/content/runs/detect/runs_colab/baseline_yolo11n/weights/best.pt")
metrics = model.val(data="VisDrone.yaml", split="val", plots=True)

print("Baseline Results")
print("mAP50-95:", metrics.box.map)
print("mAP50:", metrics.box.map50)
print("Precision:", metrics.box.mp)
print("Recall:", metrics.box.mr)

Ultralytics 8.4.38 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLO11n summary (fused): 101 layers, 2,584,102 parameters, 0 gradients, 6.3 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1598.9±759.1 MB/s, size: 126.7 KB)
val: Scanning /content/datasets/VisDrone/labels/val.cache... 548 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 548/548 114.9Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 35/35 3.1it/s 11.2s
                   all        548      38759      0.407      0.317      0.282      0.158
            pedestrian        520       8844      0.384      0.357      0.298      0.121
                people        482       5125       0.45      0.225      0.208     0.0679
               bicycle        364       1287      0.201     0.0946     0.0545      0.018
                   car        515      14064      0.577      0.745      0.708      0.465
                   van        421  

In [ ]:
import os

run_dir = "/content/runs/detect/runs_colab/baseline_yolo11n"
print("Baseline folder contents:")
for f in sorted(os.listdir(run_dir)):
    print(f)

Baseline folder contents:
BoxF1_curve.png
BoxPR_curve.png
BoxP_curve.png
BoxR_curve.png
args.yaml
confusion_matrix.png
confusion_matrix_normalized.png
labels.jpg
results.csv
results.png
train_batch0.jpg
train_batch1.jpg
train_batch2.jpg
train_batch8100.jpg
train_batch8101.jpg
train_batch8102.jpg
val_batch0_labels.jpg
val_batch0_pred.jpg
val_batch1_labels.jpg
val_batch1_pred.jpg
val_batch2_labels.jpg
val_batch2_pred.jpg
weights


In [ ]:
from ultralytics import YOLO

model = YOLO("yolo11n.pt")

model.train(
    data="VisDrone.yaml",
    epochs=12,
    imgsz=960,
    batch=8,
    device=0,
    workers=2,
    project="runs_colab",
    name="exp_imgsz960",
    save=True,
    plots=True,
    patience=5,
    verbose=True
)

Ultralytics 8.4.38 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=8, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=VisDrone.yaml, degrees=0.0, deterministic=True, device=0, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=12, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=960, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolo11n.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=exp_imgsz9602, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=5, perspective=0

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7e92d1d01c70>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.0

Experiment 1 investigates the impact of increasing input image resolution from 640 to 960.

Compared to the baseline model, the results show a clear improvement:
- mAP50-95 increased from 0.158 to 0.199
- mAP50 increased from 0.282 to 0.338
- Precision increased from 0.407 to 0.465
- Recall increased from 0.317 to 0.360

This improvement is expected because higher resolution images provide more detailed spatial information, which is particularly beneficial for detecting small objects. The VisDrone dataset contains many small and densely packed objects such as pedestrians and bicycles, making resolution an important factor.

However, increasing resolution also significantly increases training time and GPU memory usage. In this experiment, the batch size was reduced to accommodate the larger image size.

Overall, the results demonstrate that input resolution is a critical factor in object detection performance, especially for small-object detection tasks.


In [ ]:
from ultralytics import YOLO

model = YOLO("/content/runs/detect/runs_colab/exp_imgsz9602/weights/best.pt")
metrics = model.val(data="VisDrone.yaml")

print("Experiment 1 Results")
print("mAP50-95:", metrics.box.map)
print("mAP50:", metrics.box.map50)
print("Precision:", metrics.box.mp)
print("Recall:", metrics.box.mr)

Ultralytics 8.4.38 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLO11n summary (fused): 101 layers, 2,584,102 parameters, 0 gradients, 6.3 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2447.6±726.4 MB/s, size: 153.1 KB)
val: Scanning /content/datasets/VisDrone/labels/val.cache... 548 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 548/548 135.2Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 35/35 2.3it/s 15.0s
                   all        548      38759      0.464       0.36      0.338      0.199
            pedestrian        520       8844      0.487       0.41      0.388      0.175
                people        482       5125      0.549      0.262      0.278      0.101
               bicycle        364       1287      0.257      0.139     0.0922     0.0405
                   car        515      14064      0.661      0.782      0.772      0.525
                   van        421  

Experiment 1 evaluates the impact of increasing the input image resolution from 640 to 960.

Compared to the baseline model, the performance improves significantly:
- mAP50-95 increases from 0.158 to 0.199
- mAP50 increases from 0.282 to 0.338
- Precision increases from 0.407 to 0.464
- Recall increases from 0.317 to 0.360

This improvement is expected because higher resolution images provide more detailed spatial information, which is especially beneficial for detecting small objects. The VisDrone dataset contains many small and densely packed objects, making resolution a critical factor.

However, increasing the resolution also increases computational cost and memory usage. To accommodate this, the batch size was reduced from 16 to 8.

Overall, the results demonstrate that increasing input resolution is an effective strategy for improving detection performance on small-object datasets.

In [ ]:
from ultralytics import YOLO

model = YOLO("yolo11n.pt")

model.train(
    data="VisDrone.yaml",
    epochs=12,
    imgsz=640,
    batch=16,
    device=0,
    workers=2,
    project="runs_colab",
    name="exp_aug",
    hsv_h=0.03,
    hsv_s=0.9,
    hsv_v=0.5,
    degrees=5.0
)

Ultralytics 8.4.38 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=VisDrone.yaml, degrees=5.0, deterministic=True, device=0, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=12, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.03, hsv_s=0.9, hsv_v=0.5, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolo11n.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=exp_aug, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=100, perspective=0.0, 

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7e920bf7b530>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.0

Experiment 2 investigates the effect of data augmentation on model performance.

Compared to the baseline, the results show a decrease in performance:
- mAP50-95 decreased from 0.158 to 0.130
- mAP50 decreased from 0.282 to 0.238
- Precision and Recall also decreased

This indicates that the applied augmentation strategy negatively impacted the model.

One possible reason is that the VisDrone dataset contains many small objects. Strong augmentations such as high saturation changes (hsv_s=0.9) and brightness variation (hsv_v=0.5) may distort important visual features, making it harder for the model to learn.

Additionally, small objects are more sensitive to transformations like rotation, which can further degrade performance.

Overall, this experiment demonstrates that augmentation must be carefully tuned, especially for datasets with small objects.

In [ ]:
from ultralytics import YOLO

model = YOLO("/content/runs/detect/runs_colab/exp_aug/weights/best.pt")
metrics = model.val(data="VisDrone.yaml")

print("Experiment 2 Results")
print("mAP50-95:", metrics.box.map)
print("mAP50:", metrics.box.map50)

Ultralytics 8.4.38 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLO11n summary (fused): 101 layers, 2,584,102 parameters, 0 gradients, 6.3 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2254.6±995.2 MB/s, size: 153.1 KB)
val: Scanning /content/datasets/VisDrone/labels/val.cache... 548 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 548/548 229.8Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 35/35 3.2it/s 11.1s
                   all        548      38759       0.36      0.278      0.238       0.13
            pedestrian        520       8844      0.332      0.333      0.256     0.0992
                people        482       5125      0.437      0.203      0.182     0.0582
               bicycle        364       1287      0.149      0.035     0.0231    0.00726
                   car        515      14064      0.476      0.745      0.677      0.424
                   van        421  

Three experiments were conducted to evaluate different strategies for improving object detection performance on the VisDrone dataset.

The baseline model achieved an mAP50-95 of 0.158.

In Experiment 1, increasing the input image resolution from 640 to 960 resulted in a significant improvement, raising mAP50-95 to 0.199. This confirms that higher resolution is beneficial for detecting small objects, which are prevalent in the VisDrone dataset.

In Experiment 2, additional data augmentation was applied, including strong color transformations and rotation. However, this resulted in a decrease in performance, with mAP50-95 dropping to 0.130. This suggests that aggressive augmentation can distort important features, particularly for small objects.

Overall, the results indicate that resolution is a more effective factor than augmentation for this dataset. Careful tuning of augmentation strategies is necessary, especially when dealing with small-object detection tasks.

In [ ]:
from ultralytics import YOLO
import pandas as pd

models = ["yolo11n.pt", "yolov8n.pt", "yolov5nu.pt", "yolov10n.pt"]
rows = []

for m in models:
    name = m.replace(".pt", "")
    print(f"\nTraining {m}")

    model = YOLO(m)
    model.train(
        data="VisDrone.yaml",
        epochs=6,
        imgsz=640,
        batch=16,
        project="runs_colab",
        name=f"compare_{name}",
        device=0
    )

    best = YOLO(f"/content/runs/detect/runs_colab/compare_{name}/weights/best.pt")
    metrics = best.val(data="VisDrone.yaml")

    rows.append({
        "model": name,
        "mAP50-95": metrics.box.map,
        "mAP50": metrics.box.map50
    })

df = pd.DataFrame(rows)
df


Training yolo11n.pt
Ultralytics 8.4.38 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=VisDrone.yaml, degrees=0.0, deterministic=True, device=0, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=6, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolo11n.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=compare_yolo11n, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, pa

,model,mAP50-95,mAP50
0,yolo11n,0.118448,0.214257
1,yolov8n,0.116571,0.212883
2,yolov5nu,0.106211,0.192997
3,yolov10n,0.107807,0.197723


Four YOLO models (YOLO11n, YOLOv8n, YOLOv5nu, and YOLOv10n) were evaluated on the VisDrone dataset using identical training settings (6 epochs, image size 640).

Among the models, YOLO11n achieved the best performance with an mAP50-95 of 0.118 and mAP50 of 0.214. YOLOv8n performed similarly but slightly lower, while YOLOv10n and YOLOv5nu showed weaker performance.

The results indicate that YOLO11n provides the best balance between accuracy and efficiency for this dataset.